src/Domain/Orientation.hpp

```diff
@@ -0,0 +1,70 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+/// \file
+/// Defines class template Orientation.
+
+#pragma once
+
+#include <array>
+#include <iosfwd>
+#include <memory>
+
+#include "Domain/Direction.hpp"
+#include "Domain/SegmentId.hpp"
+
+/*!
+ * \ingroup DomainCreators
+ * \brief Allows one to construct a custom Orientation between two Blocks.
+ * \usage The user provides 2*VolumeDim Directions, encoding
+ * the correspondence between the directions in each block.
+ * \tparam VolumeDim the dimension of the objects.
+ */
+template <size_t VolumeDim>
+class Orientation {
+ public:
+  Orientation() = default;
+  Orientation<VolumeDim>(
+      std::array<Direction<VolumeDim>, VolumeDim> mapped_directions);
+  Orientation<VolumeDim>(
+      const std::array<Direction<VolumeDim>, VolumeDim>& directions_in_host,
+      const std::array<Direction<VolumeDim>, VolumeDim>&
+          directions_in_neighbor);
+  ~Orientation() = default;
+  Orientation(const Orientation<VolumeDim>&) = delete;
+  Orientation<VolumeDim>& operator=(const Orientation<VolumeDim>&) = delete;
+  Orientation(Orientation<VolumeDim>&& /* rhs */) noexcept = default;  // NOLINT
+  Orientation<VolumeDim>& operator=(                                   // NOLINT
+      Orientation<VolumeDim>&& /*rhs*/) noexcept = default;
+
+  /// Return a copy of the concrete Orientation.
+  std::unique_ptr<Orientation<VolumeDim>> clone() const {
+    return std::make_unique<Orientation<VolumeDim>>(mapped_directions_);
+  }
+
+  /// True only for Orientations::Aligned.
+  bool is_aligned() const noexcept { return is_aligned_; }
+
+  /// The corresponding dimension in the neighbor.
+  size_t mapped(const size_t dim) const noexcept {
+    return mapped_directions_[dim].dimension();
+  }
+
+  /// The corresponding direction in the neighbor.
+  Direction<VolumeDim> mapped(const Direction<VolumeDim>& direction) const {
+    return direction.side() == Side::Upper
+               ? mapped_directions_[direction.dimension()]
+               : mapped_directions_[direction.dimension()].opposite();
+  }
+
+  /// The corresponding SegmentIds in the neighbor.
+  std::array<SegmentId, VolumeDim> mapped(
+      const std::array<SegmentId, VolumeDim>& segmentIds) const;
+
+  /// Adds a string to the stream corresponding to the concrete orientation.
+  std::ostream& write(std::ostream& os) const;
```

Comment:
Okay, I think this should be a stream operator now since nothing inherits off of Orientation (it has no virtual functions).